src/tests/test_sexual_behaviour.py

```diff
@@ -104,3 +106,225 @@ def test_initial_sex_behaviour_groups():
+    pop.data["partnered"] = False
+    pop.data.loc[indices, "partnered"] = True
+    SBM = SexualBehaviourModule()
+    SBM.update_rred_long_term_partnered(pop.data)
+    for i in indices:
+        assert pop.data.loc[i, "rred_long_term_partnered"] == SBM.ltp_risk_factor
+
+
+def test_rred_adc():
+    N = 20
+    pop = Population(size=N, start_date=date(1989, 1, 1))
+    pop.data["HIV_status"] = False  # make test independent of how HIV is initialised
+    init_HIV_idx = [randint(0, N-1) for i in range(5)]
+    pop.data.loc[init_HIV_idx, "HIV_status"] = True
+    expected_rred = np.ones(N)
+    expected_rred[init_HIV_idx] = 0.2
+    SBM = SexualBehaviourModule()
+    # initialise rred_adc
+    SBM.init_risk_factors(pop.data)
+    # Check only people with HIV have rred_adc = 0.2
+    assert np.all(pop.data["rred_adc"] == expected_rred)
+    # Assign more people with HIV
+    add_HIV_idx = [randint(0, N-1) for i in range(5)]
+    pop.data.loc[add_HIV_idx, "HIV_status"] = True
+    # Update rred factors
+    SBM.update_sex_behaviour(pop)
+    expected_rred[add_HIV_idx] = 0.2
+    assert np.all(pop.data["rred_adc"] == expected_rred)
+
+
+def test_rred_diagnosis():
+    N = 20
+    pop = Population(size=N, start_date=date(1989, 1, 1))
+    pop.data["HIV_status"] = False  # init everyone HIV negative
+    SBM = SexualBehaviourModule()
+    SBM.rred_diagnosis = 4
+    SBM.rred_diagnosis_period = timedelta(700)
+    SBM.update_rred_diagnosis(pop.data, pop.date)
+    assert np.all(pop.data["rred_diagnosis"] == 1)
+    # give up some people HIV and advance the date
+    HIV_idx = [randint(0, N-1) for i in range(5)]
+    pop.data.loc[HIV_idx, "HIV_status"] = True
+    pop.data.loc[HIV_idx, "HIV_Diagnosis_Date"] = pop.date
+    SBM.update_rred_diagnosis(pop.data, pop.date)
+    for i in HIV_idx:
+        assert pop.data.loc[i, "rred_diagnosis"] == 4
+    pop.date += timedelta(days=365)
+    SBM.update_rred_diagnosis(pop.data, pop.date)
+    for i in HIV_idx:
+        assert pop.data.loc[i, "rred_diagnosis"] == 4
+    pop.date += timedelta(days=500)
+    SBM.update_rred_diagnosis(pop.data, pop.date)
+    for i in HIV_idx:
+        assert pop.data.loc[i, "rred_diagnosis"] == 2
+
+
+def test_rred_balance():
+    N = 1000
+    n_partners = 1000
+    pop = Population(size=N, start_date=date(1989, 1, 1))
+    SBM = SexualBehaviourModule()
+    # need to set the number of new partners to test for risk reduction cases
+    pop.data["num_partners"] = 0
+    men = selector(pop.data, sex=(operator.eq, SexType.Male))
+    women = selector(pop.data, sex=(operator.eq, SexType.Female))
+
+    # equal numbers of partners
+    n_partners_men = n_partners//2
+    n_partners_women = n_partners//2
+    # distribute partners amongst the men and women
+    men_idx = pop.data.loc[men].index
+    women_idx = pop.data.loc[women].index
+    for i in range(n_partners_men):
+        pop.data.loc[rng.choice(men_idx), "num_partners"] += 1
```

Comment:
Could also do one call to the rng. Not sure if it saves much time though. Only mention it because it seems a bit more natural to me, but not particularly tied to it!
```suggestion
    for i in rng.choice(men_idx, size=n_partners_men):
        pop.data.loc[i, "num_partners"] += 1
```
(it samples with replacement by default, so this should be the same?)